cmd/main.go

```diff
@@ -1,136 +1,173 @@
-`
-)
+func printMsg(m *pool.Message) {
+	if m != nil {
+		log.Printf("\nMESSAGE:\n%s", m.String())
+	}
+	body, err := m.ReadBody()
+	if err != nil {
+		log.Fatalf("failed to read body %v", err)
+	}
+	if len(body) > 0 {
+		log.Printf("\nMESSAGE BODY:\n%s", string(body))
+	}
+}
 
-func parseCode(code string) (codes.Code, error) {
+func makeRequest(code codes.Code, args []string) {
+	client, err := coap.New(host + ":" + port)
+	if err != nil {
+		log.Fatalf("Error coap creating client: %v", err)
+	}
+	var opts coapmsg.Options
+	if options != nil {
+		for _, optString := range options {
+			opt := strings.Split(optString, ",")
+			if len(opt) < 2 {
+				log.Fatal("Invalid option format")
+			}
+			optId, err := strconv.ParseUint(opt[0], 10, 16)
+			if err != nil {
+				log.Fatal("Error parsing option id")
+			}
+			opts = append(opts, coapmsg.Option{ID: coapmsg.OptionID(optId), Value: []byte(opt[1])})
+		}
+	}
+	if auth != "" {
+		opts = append(opts, coapmsg.Option{ID: coapmsg.URIQuery, Value: []byte("auth=" + auth)})
+	}
 	switch code {
-	case get:
-		return codes.GET, nil
-	case put:
-		return codes.PUT, nil
-	case post:
-		return codes.POST, nil
-	case delete:
-		return codes.DELETE, nil
+	case codes.GET:
+		switch {
+		case observe:
+			obs, err := client.Receive(args[0], opts...)
+			if err != nil {
+				log.Fatalf("Error observing resource: %v", err)
+			}
+			errs := make(chan error, 2)
+			go func() {
+				c := make(chan os.Signal)
+				signal.Notify(c, syscall.SIGINT)
+				errs <- fmt.Errorf("%s", <-c)
+			}()
+
+			err = <-errs
+			obs.Cancel(context.Background(), opts...)
+			log.Fatalf("Observation terminated: %v", err)
+		default:
+			res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), nil, opts...)
+			if err != nil {
+				log.Fatalf("Error sending message: %v", err)
+			}
+			printMsg(res)
+		}
+	default:
+		pld := strings.NewReader(data)
+		res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), pld, opts...)
+		if err != nil {
+			log.Fatalf("Error sending message: %v", err)
+		}
+		printMsg(res)
+
 	}
-	return 0, errors.New("Message can be GET, POST, PUT or DELETE")
 }
 
-func printMsg(m *pool.Message) {
-	if m != nil {
-		log.Printf("\nMESSAGE:\n %v", m)
+func checkArgs(cmd *cobra.Command, args []string) bool {
+	if len(args) < 1 {
+		fmt.Fprintf(os.Stdout, color.YellowString("\nusage: %s\n\n"), cmd.Use)
+		return false
 	}
+	return true
 }
 
 func main() {
-	if len(os.Args) < 2 {
-		log.Fatal(helpCmd)
-	}
-	help := strings.ToLower(os.Args[1])
-	if help == "-h" || help == "--help" {
-		log.Println(helpMsg)
-		os.Exit(0)
+	var rootCmd = &cobra.Command{
+		Use:   "coap-cli <method> <URL> [options]",
+		Short: "CLI for CoAP",
 	}
 
-	code, err := parseCode(strings.ToUpper(os.Args[1]))
-	if err != nil {
-		log.Fatalf("Can't read request code: %s\n%s", err, helpCmd)
+	var getCmd = &cobra.Command{
+		Use:   "get <url>",
+		Short: "Perform a GET request on a COAP resource",
```

Comment:
Add examples at least 2 one which short form and one in long form for all commands